### PR TITLE
Port thrust's pinned_allocator to cudf, since Thrust 1.17 removes the type

### PR DESCRIPTION
**Pull request by robertmaynard**
## Description
Thrust 1.17 removes the experimental/pinned_allocator. While Thrust offers a replacement in `thrust::system::cuda::universal_host_pinned_memory_resource`. In doing so we also need to move the consumers to being CUDA sources which would negatively impact our compile time.

Instead we move Thrust's removed pinned_allocator into cudf as it allows usage from C++ sources and doesn't
require larger changes to handle the fact the value_type from the container becomes `thrust::pointer<T>`.

Note: We haven't seen a compile failure up to this point due to the fact that all CUDA 11.X toolkits provide a version
of thrust that has the experimental header. So when it wasn't found in our 1.17.2 location the compiler would fallback
to the one in the CTK.  We can't rely on this behavior moving forward.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by bdice — APPROVED** [review]
I saw the Slack conversation about this -- all looks good to me.

**Comment by robertmaynard** [comment]
Moving this to do not merge as I missed another places that libcudf uses the experimental pinned allocator.

**Comment by davidwendt** [comment]
I did not see the Slack conversation. Why are the the .cpp files renamed? Are there any device code launches in the new header?

**Comment by robertmaynard** [comment]
> I did not see the Slack conversation. Why are the the .cpp files renamed? Are there any device code launches in the new header?

Due to the usage of `thrust::system::cuda::universal_host_pinned_memory_resource>` it requires the input files to be CUDA for it to select the correct device when it tries to deallocate or resize the thrust::host_vector. I consider this a bug / oversight in the thrust implementation.

I am not sure this is the approach we should take compared to implementing our own pinned allocator that doesn't require compilation via CUDA.

**Comment by davidwendt** [comment]
> I consider this a bug / oversight in the thrust implementation.

Is it an option to fix this in thrust and then include the fix into our code? I fear how this will spread.

**Comment by robertmaynard** [comment]
@bdice @davidwendt 

I have completly redesigned the way we fix this issue. Instead of using `universal_host_pinned_memory_resource` I have instead ported the removed `pinned_allocator` to cudf. This simplifies the using code as it doesn't need to be aware that the value type of the containers become `thrust::pointer<T>` instead of `T*`. Plus it allows us to continue keeping these source files as C++.

Once https://github.com/NVIDIA/libcudacxx/pull/309 has been merged we can explore using that instead of our custom pinned_allocator.

File: cpp/include/cudf/utilities/pinned_allocator.h
Comment (upsj) [inline cpp/include/cudf/utilities/pinned_allocator.h:53]: nit: 
```suggestion
    using value_type = void;
    using pointer = void*;
    using const_pointer = const void*;
    using size_type = std::size_t;
    using difference_type = std::ptrdiff_t;
```

**Review by upsj — APPROVED** [review]
LGTM in general, the code style in the allocator seems a bit off to me, but that's just a small nit.

File: cpp/include/cudf/utilities/pinned_allocator.h
Comment (bdice) [inline cpp/include/cudf/utilities/pinned_allocator.h:38]: This strategy looks fine overall— can the class be in a detail namespace or is it ever returned to / passed by the user? From a quick glance it seemed that detail might be fine. Just don’t want to extend our public API for this.

Comment (robertmaynard) [reply cpp/include/cudf/utilities/pinned_allocator.h:38]: Very good point, Moved the class to `thrust::detail`

File: cpp/include/cudf/utilities/pinned_allocator.h
Comment (bdice) [inline cpp/include/cudf/utilities/pinned_allocator.h:128]: This unused parameter gives Doxygen an error. Maybe remove from the docstring or write as a note.

File: cpp/include/cudf/utilities/pinned_allocator.h
Comment (bdice) [inline cpp/include/cudf/utilities/pinned_allocator.h:154]: Same issue with Doxygen as above.

**Review by bdice — APPROVED** [review]
Approving - just some minor style hiccups to fix.

File: cpp/include/cudf/utilities/pinned_allocator.hpp
Comment (jrhemstad) [inline cpp/include/cudf/utilities/pinned_allocator.hpp:27]: This file should be in `include/cudf/detail/utilities`

File: cpp/include/cudf/utilities/pinned_allocator.hpp
Comment (jrhemstad) [inline cpp/include/cudf/utilities/pinned_allocator.hpp:157]: I know this is just copy/paste from Thrust, but we should probably do a bit more than that and use things like `CUDF_CUDA_TRY` for error checking and throw `cudf::` specific exceptions.

Comment (jrhemstad) [reply cpp/include/cudf/utilities/pinned_allocator.hpp:157]: I think that would eliminate the need for this header to include `thrust` headers as well, especially `detail` headers.

File: cpp/include/cudf/utilities/pinned_allocator.hpp
Comment (robertmaynard) [reply cpp/include/cudf/utilities/pinned_allocator.hpp]: Good observation. I have moved the header and updated consumers

File: cpp/include/cudf/utilities/pinned_allocator.hpp
Comment (robertmaynard) [reply cpp/include/cudf/utilities/pinned_allocator.hpp]: Would throwing `cudf::` exceptions have any impact on this class being used by `thrust::host_vector` does it expect that all execptions throwin are derived from `thrust` exceptions?

Comment (jrhemstad) [reply cpp/include/cudf/utilities/pinned_allocator.hpp]: > expect that all execptions throwin are derived from thrust exceptions

I would certainly hope not. The `Allocator` is intended to user provided and there should be no exception to derive from anything other than the appropriate `std::` exceptions.

**Review by jrhemstad — APPROVED** [review]

**Comment by robertmaynard** [comment]
@gpucibot merge

**Comment by robertmaynard** [comment]
@gpucibot merge

**Comment by davidwendt** [comment]
Looks like you need ops approval before merging this.

**Review by msadang — APPROVED** [review]
LGTM

**Pull request merged**
